celery/canvas.py

```diff
@@ -550,86 +550,97 @@ def set(self, immutable=None, **options):
     def set_immutable(self, immutable):
         self.immutable = immutable
 
-    def stamp(self, visitor=None, **headers):
-        """Stamp this signature with additional custom headers.
-        Using a visitor will pass on responsibility for the stamping
-        to the visitor.
-
-        .. versionadded:: 5.3
-
-        Arguments:
-            visitor (StampingVisitor): Visitor API object.
-            headers (Dict): Stamps that should be added to headers.
-        """
-        self.stamp_links(visitor, **headers)
-        headers = headers.copy()
-        visitor_headers = None
-        if visitor is not None:
-            visitor_headers = visitor.on_signature(self, **headers) or {}
-        headers = self._stamp_headers(visitor_headers, **headers)
-        return self.set(**headers)
-
-    def _stamp_headers(self, visitor_headers=None, **headers):
-        """ Collect all stamps from visitor, headers and self,
+    def _stamp_headers(self, visitor_headers=None, append_stamps=True, self_headers=True, **headers):
+        """Collect all stamps from visitor, headers and self,
         and return an idempotent dictionary of stamps.
 
         .. versionadded:: 5.3
 
         Arguments:
             visitor_headers (Dict): Stamps from a visitor method.
+            append_stamps (bool):
+                If True, duplicated stamps will be appended to a list.
+                If False, duplicated stamps will be replaced by the last stamp.
+            self_headers (bool):
+                If True, stamps from self.options will be added.
+                If False, stamps from self.options will be ignored.
             headers (Dict): Stamps that should be added to headers.
 
         Returns:
             Dict: Merged stamps.
         """
-        # Use aggregate_duplicates=False to prioritize visitor_headers over headers in case of duplicated stamps.
+        # Use append_stamps=True to prioritize visitor_headers over headers in case of duplicated stamps.
         # This will lose duplicated headers from the headers argument, but that is the best effort solution
         # to avoid implicitly casting the duplicated stamp into a list of both stamps from headers and
         # visitor_headers of the same key.
         # Example:
         #   headers = {"foo": "bar1"}
         #   visitor_headers = {"foo": "bar2"}
-        #   _merge_dictionaries(headers, visitor_headers, aggregate_duplicates=True)
+        #   _merge_dictionaries(headers, visitor_headers, append_stamps=True)
         #   headers["foo"] == ["bar1", "bar2"] -> The stamp is now a list
-        #   _merge_dictionaries(headers, visitor_headers, aggregate_duplicates=False)
+        #   _merge_dictionaries(headers, visitor_headers, append_stamps=False)
```

Comment:
You haven't changed the `_merge_dictionaries` signature to match this comment.
The argument name is still `aggregate_duplicates`, see this [line](https://github.com/celery/celery/pull/8064/files#diff-3a80ff45da16a11b96e26a63973d7d490187a68ddc1949e2dfd7fd090b208841R596) for example.

Reply:
That's because `_merge_dictionaries` uses `aggregate_duplicates` and this uses `append_stamps`, which is sent to `aggregate_duplicates` of `_merge_dictionaries`.

`append_stamps` - logical meaning in the stamping context.
`aggregate_duplicates` - logical meaning when merging dicts.
They are matching when merging stamps dicts.

Reply:
You're not following me. This comment contains code that will not work when copy/pasted.

Reply:
Apologies, I understand now 🙏 
Fix pushed.